backends/p4tools/testgen/core/program_info.h

```diff
@@ -0,0 +1,109 @@
+#ifndef TESTGEN_CORE_PROGRAM_INFO_H_
+#define TESTGEN_CORE_PROGRAM_INFO_H_
+
+#include <stddef.h>
+
+#include <vector>
+
+#include "ir/ir.h"
+#include "lib/cstring.h"
+#include "lib/null.h"
+
+#include "backends/p4tools/testgen/lib/concolic.h"
+#include "backends/p4tools/testgen/lib/continuation.h"
+#include "backends/p4tools/testgen/lib/namespace_context.h"
+
+namespace P4Tools {
+
+namespace P4Testgen {
+
+/// Stores target-specific information about a P4 program.
+class ProgramInfo {
+ private:
+    const NamespaceContext* globalNameSpaceContext;
+
+ protected:
+    explicit ProgramInfo(const IR::P4Program* program);
+
+    /// The list of concolic methods implemented by the target. This list is assembled during
+    /// initialization.
+    ConcolicMethodImpls concolicMethodImpls;
+
+    std::vector<Continuation::Command> pipelineSequence;
+
+    boost::optional<const Constraint*> targetConstraints = boost::none;
+
+ public:
+    ProgramInfo(const ProgramInfo&) = default;
+
+    ProgramInfo(ProgramInfo&&) = default;
+
+    ProgramInfo& operator=(const ProgramInfo&) = default;
+
+    ProgramInfo& operator=(ProgramInfo&&) = default;
+
+    virtual ~ProgramInfo() = default;
+
+    /// The P4 program from which this object is derived.
+    const IR::P4Program* program;
+
+    /// Casts this object to T. Useful for casting to a concrete implementation of this class.
```

Comment:
this is repeated multiple times, please move it to a base abstract class

Reply:
Should this be a general abstract class for the compiler?

Reply:
I think we could have only one for the whole compiler, probably in lib.